pkg/engine/diff.go

```diff
@@ -230,7 +230,25 @@ func printObject(
 // GetResourceOutputsPropertiesString prints only those properties that either differ from the input properties or, if
 // there is an old snapshot of the resource, differ from the prior old snapshot's output properties.
 func GetResourceOutputsPropertiesString(
-	step StepEventMetadata, indent int, planning bool, debug bool) string {
+	step StepEventMetadata, indent int, planning bool, debug bool, refresh bool) string {
+	// We should only attempt to write out outputs if we are sure that this resource's outputs are complete. We know
+	// this for sure when we are performing updates, since we have just performed some resource operation and received
+	// outputs, but for previews we must be careful to not display a diff against outputs that are not complete.
+	if planning {
+		// If we are doing a preview, the only times that we are confident that our output bag is complete is when we
+		// are doing a refresh. So when we aren't...
+		if !refresh {
+			// ... we can only be sure that our outputs are complete if we have performed a Read or ReadReplacement, in
+			// which case the provider has given us a complete set of outputs for this resource.
+			if step.Op != deploy.OpRead && step.Op != deploy.OpReadReplacement {
+				return ""
```

Comment:
Clearly @CyrusNajmabadi and I have different styles, as I find this and its associated rightwards growth borderline unreadable ;)